scio-core/src/main/java/com/spotify/scio/transforms/BatchedBaseAsyncLookupDoFn.java

```diff
@@ -0,0 +1,330 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.spotify.scio.transforms;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.cache.Cache;
+import com.spotify.scio.transforms.BaseAsyncLookupDoFn.CacheSupplier;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Semaphore;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.commons.lang3.tuple.Triple;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link DoFn} that performs asynchronous lookup using the provided client. Lookup requests may
+ * be deduplicated.
+ *
+ * @param <Input> input element type.
+ * @param <BatchRequest> batched input element type
+ * @param <BatchResponse> batched output element type
+ * @param <Output> client lookup value type.
+ * @param <ClientType> client type.
+ * @param <FutureType> future type.
+ * @param <TryWrapper> client lookup value type wrapped in a Try.
+ */
+public abstract class BatchedBaseAsyncLookupDoFn<
+        Input, BatchRequest, BatchResponse, Output, ClientType, FutureType, TryWrapper>
+    extends DoFnWithResource<Input, KV<Input, TryWrapper>, Pair<ClientType, Cache<String, Output>>>
+    implements FutureHandlers.Base<FutureType, BatchResponse> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BatchedBaseAsyncLookupDoFn.class);
+
+  // Data structures for handling async requests
+  private final int batchSize;
+  private final SerializableFunction<List<Input>, BatchRequest> batchRequestFn;
+  private final SerializableFunction<BatchResponse, List<Pair<String, Output>>> batchResponseFn;
+  private final SerializableFunction<Input, String> idExtractorFn;
+  private final int maxPendingRequests;
+  private final CacheSupplier<String, Output> cacheSupplier;
+
+  private final Semaphore semaphore;
+  private final ConcurrentMap<UUID, FutureType> futures = new ConcurrentHashMap<>();
+  private final ConcurrentMap<String, List<Triple<Input, Instant, BoundedWindow>>> inputs =
+      new ConcurrentHashMap<>();
+
+  private final Queue<Input> batch = new ArrayDeque<>();
```

Comment:
No need of a concurrent collection here as it is only accessed in `processFunction` and `finishBundle` which are thread safe.

Reply:
didn't know that :)